### PR TITLE
Implement asgardeo auth provider

### PR DESCRIPTION
**Pull request by yathindrak**
## Reasoning 💡

* Add [Asgardeo](https://wso2.com/asgardeo/docs/guides/authentication/) Auth provider
* Add docs in https://github.com/nextauthjs/docs/pull/205
* Executed tests

## Checklist 🧢

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

**Comment by yathindrakodithuwakku** [comment]
Here is a proof for the authentication integration.
Screenshot 2022-01-19 at 22 51 24

File: app/pages/api/auth/[...nextauth].ts [hunk 1]
```diff
@@ -189,7 +190,14 @@ export const authOptions: NextAuthOptions = {
     PatreonProvider({
       clientId: process.env.PATREON_ID,
       clientSecret: process.env.PATREON_SECRET,
-    })
+    }),
+    AsgardeoProvider({
+      domain: process.env.ASGARDEO_DOMAIN,
+      tenantId: process.env.ASGARDEO_TENANT,
```
Comment (mefarazath) [inline app/pages/api/auth/[...nextauth].ts:196]: Is there any reason why we need both domain and tenant?

Comment (mefarazath) [reply app/pages/api/auth/[...nextauth].ts:196]: also shouldn't tenantId --> organizationName?

File: src/providers/asgardeo.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+): OAuthConfig<P> {
+  return {
+    id: "asgardeo",
+    name: "Asgardeo",
+    clientId: options?.clientId,
+    clientSecret: options?.clientSecret,
+    type: "oauth",
+    wellKnown: `https://${options.domain}/t/${options?.tenantId}/oauth2/token/.well-known/openid-configuration`,
+    authorization: {
+      params: { scope: options?.scopes || "openid email profile internal_login" }
```
Comment (mefarazath) [inline src/providers/asgardeo.ts:18]: Do we need `internal_login`?

**Comment by balazsorban44** [comment]
Good points made by @mefarazath, could you answer them @yathindrakodithuwakku?

Also, for proof of authentication, we would need an actual login proof (so after login), not just the auth page you are redirected to.

File: app/pages/api/auth/[...nextauth].ts (hunk 1 above)
Comment (yathindrakodithuwakku) [reply app/pages/api/auth/[...nextauth].ts:196]: Hi @mefarazath, Yes can proceed with organization name and doesnt require domain.
Sry for the delay :)

File: src/providers/asgardeo.ts (hunk 2 above)
Comment (yathindrakodithuwakku) [reply src/providers/asgardeo.ts]: Yes we can omit that by using userinfo route. Updated the PR

**Comment by yathindrakodithuwakku** [comment]
@balazsorban44 Please find the video on login and sign out flows as a proof.

https://user-images.githubusercontent.com/32919513/152634310-6d577a34-cf1b-4b60-b9c1-c40430273c72.mp4

**Comment by yathindrakodithuwakku** [comment]
@balazsorban44 Please find the screencast after taken fixes of merge conflicts and code simplifications.

https://user-images.githubusercontent.com/32919513/158668715-ec83c248-8b94-4001-ab1a-e7a17b8ecaee.mp4

**Review by mefarazath — APPROVED** [review]
LGTM!

File: apps/dev/pages/api/auth/[...nextauth].ts [hunk 3]
```diff
@@ -193,6 +194,12 @@ export const authOptions: NextAuthOptions = {
       clientId: process.env.PATREON_ID,
       clientSecret: process.env.PATREON_SECRET,
     }),
+    AsgardeoProvider({
+      organization: process.env.ASGARDEO_ORGANIZATION,
+      clientId: process.env.ASGARDEO_CLIENT_ID,
+      clientSecret: process.env.ASGARDEO_CLIENT_SECRET,
+      scopes: process.env.ASGARDEO_SCOPES,
```
Comment (balazsorban44) [inline apps/dev/pages/api/auth/[...nextauth].ts:201]:
```suggestion
```

I don't think we should introduce new options, it would be confusing since other providers don't have this.

File: packages/next-auth/src/providers/asgardeo.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    scopes?: string
+  }
+): OAuthConfig<P> {
+  return {
+    id: "asgardeo",
+    name: "Asgardeo",
+    type: "oauth",
+    wellKnown: `https://api.asgardeo.io/t/${options?.organization}/oauth2/token/.well-known/openid-configuration`,
+    authorization: {
+      params: { scope: options?.scopes || "openid email profile" }
```
Comment (balazsorban44) [inline packages/next-auth/src/providers/asgardeo.ts:21]:
```suggestion
      params: { scope: "openid email profile" } 
```

File: packages/next-auth/src/providers/asgardeo.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+): OAuthConfig<P> {
+  return {
+    id: "asgardeo",
+    name: "Asgardeo",
+    type: "oauth",
+    wellKnown: `https://api.asgardeo.io/t/${options?.organization}/oauth2/token/.well-known/openid-configuration`,
+    authorization: {
+      params: { scope: options?.scopes || "openid email profile" } 
+    },
+    idToken: true,
```
Comment (balazsorban44) [inline packages/next-auth/src/providers/asgardeo.ts:23]:
```suggestion
```

This is detected automatically now (whenever there is a "openid" scope present)

File: packages/next-auth/src/providers/asgardeo.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    checks: ["pkce", "state"],
+    profile(profile) {
+      return {
+        id: profile?.sub,
+        name: profile?.given_name,
+        email: profile?.email
```
Comment (balazsorban44) [inline packages/next-auth/src/providers/asgardeo.ts:29]:
```suggestion
        email: profile?.email,
        image: null
```

To be consistent with other providers. Could the provider potentially return a profile image?

File: apps/dev/pages/api/auth/[...nextauth].ts [hunk 7]
```diff
@@ -193,6 +194,12 @@ export const authOptions: NextAuthOptions = {
       clientId: process.env.PATREON_ID,
       clientSecret: process.env.PATREON_SECRET,
     }),
+    AsgardeoProvider({
+      organization: process.env.ASGARDEO_ORGANIZATION,
```
Comment (balazsorban44) [inline apps/dev/pages/api/auth/[...nextauth].ts:198]:
```suggestion
      issuer: process.env.ASGARDEO_ISSUER,
```

Let's use `issuer` which is according to the spec, and consistent with the other providers

File: packages/next-auth/src/providers/asgardeo.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  options: OAuthUserConfig<P> & {
+    organization: string
+    scopes?: string
+  }
+): OAuthConfig<P> {
+  return {
+    id: "asgardeo",
+    name: "Asgardeo",
+    type: "oauth",
+    wellKnown: `https://api.asgardeo.io/t/${options?.organization}/oauth2/token/.well-known/openid-configuration`,
```
Comment (balazsorban44) [inline packages/next-auth/src/providers/asgardeo.ts:19]:
```suggestion
    wellKnown: `${options.issuer}/oauth2/token/.well-known/openid-configuration`,
```

I'm not really sure why the well-known endpoint is under /oauth2/token, but this seems to be correct after what I could find in:

https://github.com/asgardeo/asgardeo-auth-spa-sdk/blob/6d11d2627d0fe6bd3c1dc48b327b6c5dc42c1a62/README.md#the-authclientconfig-interface

You could potentially take it up with them, since I think this is not spec-compliant. According to the spec:

> OpenID Providers supporting Discovery MUST make a JSON document available at the path formed by concatenating the string /.well-known/openid-configuration to the Issuer

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig1

So I think this should be:
`${options.issuer}/.well-known/openid-configuration`

Where `options.issuer` is `https://api.asgardeo.io/t/organization`

**Review by balazsorban44 — DISMISSED** [review]
Thanks, added some comments!

**Comment by balazsorban44** [comment]
Could we also add a page to the documentation under: https://github.com/nextauthjs/next-auth/tree/main/docs/docs/providers

File: packages/next-auth/src/providers/asgardeo.ts (hunk 6 above)
Comment (yathindrakodithuwakku) [reply packages/next-auth/src/providers/asgardeo.ts]: Added the image parameter under profile settings

File: apps/dev/pages/api/auth/[...nextauth].ts (hunk 7 above)
Comment (yathindrakodithuwakku) [reply apps/dev/pages/api/auth/[...nextauth].ts]: +1

**Comment by yathindrakodithuwakku** [comment]
@balazsorban44 Thanks a lot for the suggestions. I made the improvements in the latest commit. Could you please check?

Please find the updated documentation pr as well: https://github.com/nextauthjs/docs/pull/205

Also, please find the screencast with the updated changes as the evidence. 
https://user-images.githubusercontent.com/32919513/167654811-2d218ead-d5b2-49e5-acdb-38e2bd2315bd.mp4

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]
Thanks! I added another comment. Besides, please include the docs change in this PR also. See: https://github.com/nextauthjs/docs/pull/205#issuecomment-1143253010

File: packages/next-auth/src/providers/asgardeo.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+export interface AsgardeoProfile {
+  sub: string
+  given_name: string
+  email: string
+  picture: string
+}
+
+export default function Asgardeo<P extends Record<string, any> = AsgardeoProfile>(
+  options: OAuthUserConfig<P> & {
+    issuer: string
```
[the remark this answers is not in the document]
Comment (yathindrakodithuwakku) [reply packages/next-auth/src/providers/asgardeo.ts]: Fixed. Removed the issuer param.

**Comment by yathindrakodithuwakku** [comment]
@ThangHuuVu Removed the issuer and added the docs to align with the monorepo architecture.

File: docs/docs/providers/asgardeo.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+- After registring the application, go to protocol tab.
+- Check `code` grant type.
+- Add Authorized redirect URLs & Allowed origins fields.
+- Make Email, First Name, Photo URL user attributes mandatory from the console.
+
+In `.env` add the following entries:
+
+```
+NEXTAUTH_URL=<Your application url e.g: http://localhost:3000>
+SECRET=<Your secret here>
```
Comment (ThangHuuVu) [inline docs/docs/providers/asgardeo.md:25]: Let's remove these as they are not directly related to how to set-up the provider

File: docs/docs/providers/asgardeo.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  AsgardeoProvider({
+    clientId: process.env.ASGARDEO_CLIENT_ID,
+    clientSecret: process.env.ASGARDEO_CLIENT_SECRET,
+    issuer: process.env.ASGARDEO_ISSUER
+  }),
+],
+secret: process.env.SECRET,
+session: {
+  strategy: "jwt",
+},
```
Comment (ThangHuuVu) [inline docs/docs/providers/asgardeo.md:46]: Please remove these configs as well 🙏

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]
Hey @yathindrakodithuwakku , I think this should be ready now, please address some remaining comments on the docs and I think we're good to merge!

File: docs/docs/providers/asgardeo.md (hunk 10 above)
Comment (yathindrakodithuwakku) [reply docs/docs/providers/asgardeo.md]: Removed these env configs from the doc

**Comment by yathindrakodithuwakku** [comment]
@ThangHuuVu  The docs updates have been done. Will you be able to check and merge?

**Review by ThangHuuVu — APPROVED** [review]
LGTM!

**Comment by ThangHuuVu** [comment]
Oh sorry, one more thing, could you help add the provider to this page https://github.com/nextauthjs/next-auth/blob/fix/infer-gss-return-type/.github/ISSUE_TEMPLATE/2_bug_provider.yml under the provider type selection 🙌

File: apps/dev/pages/api/auth/[...nextauth].ts [hunk 12]
```diff
@@ -3,6 +3,7 @@ import type { NextAuthOptions } from "next-auth"
 
 // Providers
 import Apple from "next-auth/providers/apple"
+import AsgardeoProvider from "next-auth/providers/asgardeo"
```
Comment (mefarazath) [inline apps/dev/pages/api/auth/[...nextauth].ts:6]: Can't we rename this to Asgardeo?

Comment (mefarazath) [reply apps/dev/pages/api/auth/[...nextauth].ts:6]: Then it will be consistent with the display name everywhere

Comment (yathindrak) [reply apps/dev/pages/api/auth/[...nextauth].ts]: This is fixed with the latest commit

**Comment by yathindrak** [comment]
@ThangHuuVu The suggested improvements are made now  🙌

**Comment by pavinduLakshan** [comment]
Hi @ThangHuuVu , could you please let us know when this (Asgardeo Integration) will be available in next-auth?
Thanks

**Comment by ThangHuuVu** [comment]
Hi all, sorry for the long delay, I cleaned up the PR and move the code to the `core` package instead, thanks @yathindrak and everyone 🙌

**Comment by yathindrak** [comment]
Great to hear the changes are onboarded 🎉. Thanks a lot, @ThangHuuVu and everyone.
